# Лабораторная работа №3
#### Студент: Кузьмин Дмитрий
#### Группа 9382
#### Вариант 9

### 1. Диофантово уравнение

$ 2576x - 2461y = -115$ 

In [8]:
from sympy.solvers.diophantine.diophantine import base_solution_linear
from sympy.abc import k
from sympy import symbols


In [9]:
base_solution_linear(-115, 2576, -2461, k)

(107*k - 215, 112*k - 225)

Ответ: $x = -215 + 107k, y = -225 + 112k$

### 2. RSA

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 9, m = 34).
Шифрованное сообщение имеет вид (19; 21; 26; 2). Подберите закрытую часть ключа и прочитайте исходное слово.
Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “e” и “ъ”).

In [10]:
from sympy.crypto.crypto import encipher_rsa, decipher_rsa
from sympy.crypto.crypto import rsa_public_key, rsa_private_key
from sympy.core.numbers import mod_inverse


In [11]:
import math
def factorized(n):
    factors = [1]
    for t in range(2, (math.ceil((n / 2) + 1))):
        if n % t == 0:
            factors.append(t)

    factors.append(n)
    return factors[1: -1]


In [12]:
n = 34
e = 9
word = [19, 21, 26, 2]
p, q = factorized(n)
print(p, q)
decrypted = []
d = mod_inverse(e, (p - 1)*(q - 1))
print(d)
for l in word:
    decrypted.append(decipher_rsa(l, (n, d)))
cyrillic = [chr(i) for i in range(ord('а'), ord('я') + 1)]
cyrillic.remove('ъ')
print(decrypted)
for dl in decrypted:
    print(cyrillic[dl - 2], sep="", end="")

2 17
9
[19, 21, 26, 2]
суша

Ответ: СУША

## 3. Преобразование булевой функции

$ f (x, y, z) = ((zy) ∨ (x ∨ z)) \ xor \ ((x \ xor \ y)y) $

### ДНФ

In [13]:
from sympy.logic.boolalg import to_dnf, to_cnf
x, y, z = symbols('x y z')
formula = ((z&y) | (x|z)) ^ ((x ^ y) & y) 
to_dnf(formula, simplify=True)

x | (y & ~z) | (z & ~y)

### КНФ

In [14]:
to_cnf(formula, simplify=True)

(x | y | z) & (x | ~y | ~z)